File: cpp/04_DLinkedList(head).cpp
/*
	이중 연결 리스트: 알고리즘 구현(CPP)
		파일명: DLinkedList(head).cpp
			- 클래스: DNode
			- 클래스: DLinkedList
				생성자와 소멸자			: DLinkedList, ~DLinkedList
				노드 확인				: isEmpty, countNode
				노드 탐색				: frontNode, rearNode
				노드 삽입.삭제			: addRear, removeFront
				전체 원소(노드) 출력	: printLinkedList
*/

#include <iostream>
#include "DLinkedList(head).h"		// DLinkedList >> head
// #include "LinkedNode.h"			// DNode
using namespace std;

// 생성자: head_, tail_, count_
DLinkedList::DLinkedList(void)
	: head_(nullptr) { }

// 소멸자: 전체 노드 삭제
DLinkedList::~DLinkedList(void) {
	// while (!isEmpty()) removeFront();
	DNode* tNode = head_;
	while (tNode) {
		head_ = tNode->Rlink_;
		delete tNode;
		tNode = head_;
	}
}

// 빈 리스트 여부 판단
bool DLinkedList::isEmpty(void) const {
	return head_ == nullptr;
}

// 탐색: 노드의 총 개수(count)
int	DLinkedList::countNode(void) const {
	// if (isEmpty()) return 0;
	int		count = 0;
	DNode* rNode = head_;
	while (rNode) {		// while (rNode != nullptr)
		count++;
		rNode = rNode->Rlink_;
	}
	return count;
}

// 탐색: 첫 번째 노드(head)
DNode* DLinkedList::frontNode(void) const {
	// if (isEmpty()) return nullptr;
	return head_;
}

// 탐색: 맨 마지막 노드
DNode* DLinkedList::rearNode(void) const {
	// if (isEmpty()) return nullptr;
	DNode* rNode = head_;
	while (rNode->Rlink_)
		rNode = rNode->Rlink_;
	return rNode;
}

// 삽입: 맨 마지막 노드
void DLinkedList::addRear(const int& e) {
	DNode* nNode = new DNode(e);
	if (isEmpty()) head_ = nNode;
	else {
		DNode* rNode = rearNode();
		rNode->Rlink_ = nNode;
		nNode->Llink_ = rNode;
	}
}

// 삭제: 첫 번째 노드(head)
void DLinkedList::removeFront(void) {
	if (isEmpty()) return;
	DNode* old = head_;
	head_ = old->Rlink_;
	if (head_ != nullptr)
		old->Llink_ = nullptr;
	delete old;
}

// 출력(순방향): 리스트의 전체 원소(노드) 출력
void DLinkedList::printLinkedList(void) const {
	if (isEmpty()) {
		cout << "\n입력된 데이터가 없습니다..." << endl;
		return;
	}

	cout << "\n\t### 입력된 데이터: 순방향 ###" << endl;
	DNode* temp = head_;
	while (temp) {
		cout.width(3);
		cout << temp->data_ << " ->>";
		temp = temp->Rlink_;
	}
	cout << " NULL" << endl;
}

// 출력(역방향): 리스트의 전체 원소(노드) 출력
void DLinkedList::printRevLinkedList(void) const {
	if (isEmpty()) {
		cout << "\n입력된 데이터가 없습니다..." << endl;
		return;
	}

	cout << "\n\t### 입력된 데이터: 역방향 ###" << endl;
	DNode* temp = rearNode();
	while (temp) {
		cout.width(3);
		cout << temp->data_ << " ->>";
		temp = temp->Llink_;
	}
	cout << " NULL" << endl;
}

File: python/04_DLinkedList(head).py
'''
	이중 연결 리스트: 알고리즘 구현(Python)
		파일명: DLinkedList(head).py
			- 클래스: DNode
			- 클래스: DLinkedList
				생성자와 소멸자         : __init__, __del__
				노드 확인               : isEmpty, countNode
				노드 탐색               : frontNode, rearNode
				노드 삽입.삭제          : addRear, removeFront
				전체 원소(노드) 출력    : printLinkedList
'''

# 노드(DNode): 데이터(data), 링크(Llink, Rlink)
class DNode:
    def __init__(self, data):
        self.__data = data
        self.__Llink = None
        self.__Rlink = None
    def getData(self):          return self.__data      # getattr(DNode, '__data')
    def getLlink(self):         return self.__Llink     # getattr(DNode, '__Llink')
    def getRlink(self):         return self.__Rlink     # getattr(DNode, '__Rlink')
    def setData(self, data):    self.__data = data      # setattr(DNode, '__data', 'data')
    def setLlink(self, Llink):  self.__Llink = Llink    # setattr(DNode, '__Llink', 'Llink')
    def setRlink(self, Rlink):  self.__Rlink = Rlink    # setattr(DNode, '__Rlink', 'Rlink')

# 이중 연결 리스트: DLinkedList(head)
class DLinkedList:
    def __init__(self):          # 생성자
        self.__head = None       # 첫 번째 노드
        # self.__tail = None     # 맨 마지막 노드
        # self.__count = 0       # 노드의 총 개수

    # 소멸자: 전체 노드 삭제
    def __del__(self):
        # while not self.isEmpty() :
        #     self.removeFront()
        while not self.isEmpty():
            old = self.__head
            self.__head = old.getRlink()
            del old

    # 빈 리스트 여부 판단
    def isEmpty(self) -> bool:
        return self.__head == None

    # 탐색: 노드의 총 개수(count)
    def countNode(self) -> int:
        if self.isEmpty():
            return 0
        
        count = 0
        rNode = self.__head
        while rNode:
            count +=1
            rNode = rNode.getRlink()
        return count

    # 탐색: 첫 번째 노드(head)
    def frontNode(self) -> DNode:
        # if self.isEmpty():
        #    return None
        return self.__head

    # 탐색: 맨 마지막 노드
    def rearNode(self) -> DNode:
        # if self.isEmpty():
        #    return None
        rNode = self.__head
        while rNode.getRlink() :
            rNode = rNode.getRlink()
        return rNode

    # 삽입: 맨 마지막 노드
    def addRear(self, num):
        newNode = DNode(num)
        if self.isEmpty():
            self.__head = newNode
        else:
            rNode = self.rearNode()
            rNode.setRlink(newNode)
            newNode.setLlink(rNode)
            
    # 삭제: 첫 번째 노드(head)
    def removeFront(self) -> None:
        if self.isEmpty():
            return
        
        old = self.__head
        self.__head = old.getRlink()
        if self.__head:
            self.__head.setLlink(None)

        del old

    # 출력(순방향): 리스트의 전체 원소(노드) 출력
    def printLinkedList(self) -> None:
        if self.isEmpty():
            print('입력된 데이터가 없습니다...')
            return
        
        print(f'\n### 입력된 데이터(순방향): {self.countNode()} ###')
        temp = self.__head
        while temp:
            print(f'{temp.getData()} ->>', end=' ')
            temp = temp.getRlink()
        print(' None')
        
    # 출력(역방향): 리스트의 전체 원소(노드) 출력
    def printRevLinkedList(self) -> None:
        if self.isEmpty():
            print('입력된 데이터가 없습니다...')
            return
        
        print(f'\n### 입력된 데이터(역방향): {self.countNode()} ###')
        temp = self.rearNode()
        while temp :
            print(f'{temp.getData()} ->>', end=' ')
            temp = temp.getLlink()
        print(' None')

if __name__ == '__main__':
    # __head = None
    dList = DLinkedList()

    while (True):
        num = int(input('임의의 정수 입력(종료: 0): '))
        if num == 0 :
            break
        # 맨 마지막 노드로 삽입
        dList.addRear(num)

	# 전체 원소 출력
    dList.printLinkedList()         # 순방향 출력
    dList.printRevLinkedList()      # 역방향 출력

    # del dList     # dList.__del__()
